src/utransport.rs

```diff
@@ -534,121 +446,117 @@ impl UTransport for UPClientZenoh {
         }
     }
 
-    async fn receive(&self, _topic: UUri) -> Result<UMessage, UStatus> {
+    async fn receive(
+        &self,
+        _source_filter: &UUri,
+        _sink_filter: Option<&UUri>,
+    ) -> Result<UMessage, UStatus> {
         let msg = "Not implemented".to_string();
         log::error!("{msg}");
         Err(UStatus::fail_with_code(UCode::UNIMPLEMENTED, msg))
     }
 
     async fn register_listener(
         &self,
-        topic: UUri,
+        source_filter: &UUri,
+        sink_filter: Option<&UUri>,
         listener: Arc<dyn UListener>,
     ) -> Result<(), UStatus> {
-        if topic.authority.is_some() && topic.entity.is_none() && topic.resource.is_none() {
-            // This is special UUri which means we need to register for all of Publish, Notification, Request, and Response
-            // RPC response
-            self.register_response_listener(&topic, listener.clone());
-            // RPC request
-            self.register_request_listener(&topic, listener.clone())
+        let flag = UPClientZenoh::get_listener_message_type(source_filter, sink_filter)?;
+        // Publish & Notification
+        if flag.contains(MessageFlag::Publish) || flag.contains(MessageFlag::Notification) {
```

Comment:
So I think based on how I am reading this, the intention looks to be that the source and sink filters may result in more than one bit being set in the bitmask enum. That seems to be the intent. And that makes sense.

I'm having a hard time seeing how the function get_listener_message_type() can set multiple bits. It looks like a single value is set based on the if conditionals and returned. Hmm. Perhaps I need to read more 🤔

Reply:
You're right... I made the silly mistakes.
I updated how `get_listener_message_type` works and added more tests case to avoid the mistakes in the future.